### PR TITLE
mgr/dashboard: add service management for mgmt-gateway

### PR DESCRIPTION
**Pull request by Pegonzal**
Adds service management support for the mgmt-gateway.

[screen-capture (23).webm](https://github.com/user-attachments/assets/75100ff1-e3fc-44db-8156-0d29049c1901)

Video's spec file

```[ceph: root@ceph-node-00 /]# ceph orch ls --service_name mgmt-gateway --export
service_type: mgmt-gateway
service_name: mgmt-gateway
placement:
  count: 1
spec:
  ssl_ciphers:
  - ECDHE-ECDSA-AES128-GCM-SHA256
  - ECDHE-RSA-AES128-GCM-SHA256
```

Fixes: https://tracker.ceph.com/issues/66963

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by cloudbehl** [comment]
Lets have default values/selection for
1. Count
2. SSL Protocols
3. SSL Ciphers

Also add  description for what authentication means

**Comment by Pegonzal** [comment]
> Lets have default values/selection for
> 
> 1. Count

Adding one as default

> 2. SSL Protocols
> 3. SSL Ciphers

I don't think any default value here is desired 

> Also add description for what authentication means

I'll look into what description to add for it

**Comment by cloudbehl** [comment]
There should be some recommendations for user.

Like in case of SSL protocols can be check `TLSv1.2` and `TLSv1.3` in options and likewise whats the default supported for SSL Ciphers? @rkachach any idea?

**Comment by rkachach** [comment]
> > Lets have default values/selection for
> > 
> > 1. Count
> 
> Adding one as default
> 
> > 2. SSL Protocols
> > 3. SSL Ciphers
> 
> I don't think any default value here is desired
> 
> > Also add description for what authentication means
> 
> I'll look into what description to add for it

As for the count right now we can only have one instance (no HA is supported in the first release). For the configuration we support whatever nginx supports, by default we use `TLSv1.3` for strict security ad `TLSv1.2` in some ciphers combinations could be weak. The configuration we are using has been generated by `https://ssl-config.mozilla.org/#server=nginx`

**Review by afreen23 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts]: Updated it as per your comment, @afreen23

**Comment by Pegonzal** [comment]
jenkins test make check

**Comment by Pegonzal** [comment]
jenkins test dashboard

**Comment by Pegonzal** [comment]
jenkins test dashboard cephadm

**Review by afreen23 — CHANGES_REQUESTED** [review]
Hello,

Here are few suggestions from my end:

1. Fixing misalignment
image
Hey looks like the labels and checkboxes are quite off.
We need them to be aligned properly.
if you need help you may check other forms' html/css.

2. Having everything under SSL checkbox

image
Selecting SSL protocol/cipher and then choosing one more SSL check box for cert and key looks confusing so lets put protocol and cipher too with certs and key.

So just one checkbox for SSL which will open up all these options.


3. Multi-select instead of checkboxes

We can consider using [carbon multi-select here](https://web-components.carbondesignsystem.com/?path=/story/components-multi-select--default&globals=outline:true) for protocol and cipher

Thanks for the work @Pegonzal

**Comment by afreen23** [comment]
> > > Lets have default values/selection for
> > > 
> > > 1. Count
> > 
> > 
> > Adding one as default
> > > 2. SSL Protocols
> > > 3. SSL Ciphers
> > 
> > 
> > I don't think any default value here is desired
> > > Also add description for what authentication means
> > 
> > 
> > I'll look into what description to add for it
> 
> As for the count right now we can only have one instance (no HA is supported in the first release). For the configuration we support whatever nginx supports, by default we use `TLSv1.3` for strict security ad `TLSv1.2` in some ciphers combinations could be weak. The configuration we are using as default has been generated by `https://ssl-config.mozilla.org/#server=nginx` + some other tweaks.
> 
> Our default config must be secure. We should add a pop-up/warning that changing security configuration of the mgmt-gateway is `user's responsibility` and could lead to a weaker security config.

Then based on https://github.com/ceph/ceph/blob/main/src/pybind/mgr/cephadm/templates/services/mgmt-gateway/internal_server.conf.j2 we can have have some items preselected/checked.

And agree we can add a warning "Modifying security settings may weaken security" or a better wording.

Thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:24]: With `all other services` I meant, grafana, prometheus, alertmanager...
The service mgmt-gateway acts as a reverse proxy providing a single access point, meaning that these services will be available to the user from `{{currentURL}}:{{port}}/service_name`

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:933]: Changed the text

**Comment by Pegonzal** [comment]
Thanks for the review @Afreen, I tried to address all your comments. Please take a look and let me know what you think

**Comment by Pegonzal** [comment]
jenkins test make check

**Comment by Pegonzal** [comment]
jenkins test dashboard

**Comment by Pegonzal** [comment]
jenkins test dashboard cephadm

**Comment by afreen23** [comment]
Hey,

Commenting here as well just for record.

image
we need to enable this warning if user changes defaults for SSL based settings
I would assume we dont need to provide for now SSL ciphers as there are various options and it might be confusing.
Lets take the feedback and if required add then.

Thanks, great work!

**Comment by Pegonzal** [comment]
>we need to enable this warning if user changes defaults for SSL based settings

Changed it to show warning on user form changes

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:42]: Just to clarify, I believe this is the list of ciphers that will be overwritten 
https://github.com/ceph/ceph/blob/main/src/pybind/mgr/cephadm/templates/services/mgmt-gateway/external_server.conf.j2

We could expand the cipher options in the dropdown or just let the user write the desired one, I guess that would be preferred.
The main issue I see with this is the lack of validation of the ciphers. At least I can't find it if there is any in place.
We could perform frontend validation of the ciphers, but I am not sure which ones we should be taking into account.
Anyways, the recommended use of this service is to have the default configuration as it already tries to provide the strongest security options available. Therefore I just went with this three options I saw in a previous mgmt-gateway spec file example, and we can always expand on them.

What do you think @afreen23 ?

Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:42]: Lets add a validation as per the format mentioned  here: https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_ciphers
The list of cipher suites is overwhelming therefore better to have regex and string input only. Thanks @Pegonzal

Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts]: I've included the possible ciphers, and you can either type or select any of them thanks to the carbon component, you can find the updated video in the PR description, plesae @afreen23 let me know what you think.

The list of ciphers come from the previously mentioned `external_server.conf.j2`

**Comment by Pegonzal** [comment]
jenkins test make check

**Comment by Pegonzal** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/142995/
```
No identified problem
No problems were identified. If you know why this problem occurred, please add a suitable Cause for it.
```

Retriggering

**Comment by Pegonzal** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts]: Sounds good, then @Pegonzal , let me check

**Comment by rkachach** [comment]
While testing this PR I saw that the spec generated when using default config has the "port" field set to 443:

```
ceph orch ls --service_name mgmt-gateway --export
service_type: mgmt-gateway
service_name: mgmt-gateway
placement:
  hosts:
  - ceph-node-1
spec:
  port: 443
  ssl_protocols:
  - TLSv1.3 
```

This causes issues when generating the `redirect_uri`. I'd suggest by now to omit the port if it's the default one (443 for https and 80 for http). I'll change the cephadm code to handle this better latter, but for now I think we should adjust the dashboard side as all cephadm PRs have been already merged.

**Comment by Pegonzal** [comment]
> @Pegonzal @afreen23 while testing this PR I saw that the spec generated when using default config has the `port` field set to 443:
> 
> ```
> ceph orch ls --service_name mgmt-gateway --export
> service_type: mgmt-gateway
> service_name: mgmt-gateway
> placement:
>   hosts:
>   - ceph-node-1
> spec:
>   port: 443
>   ssl_protocols:
>   - TLSv1.3 
> ```
> 
> This causes issues when generating the `redirect_uri`. I'd suggest by now to omit the port if it's the default one (443 for https and 80 for http). I'll change the cephadm code to handle this better latter, but for now I think we should adjust the dashboard side as all cephadm PRs have been already merged.

Thanks @rkachach , I have updated the PR.
Now, for default:

```
[ceph: root@ceph-node-00 /]# ceph orch ls --service_name mgmt-gateway --export
service_type: mgmt-gateway
service_name: mgmt-gateway
placement:
  count: 1
spec:
  ssl_ciphers:
  - ECDSA
  - AES128
  ssl_protocols:
  - TLSv1.3
```

Other ports:
```
[ceph: root@ceph-node-00 /]# ceph orch ls --service_name mgmt-gateway --export
service_type: mgmt-gateway
service_name: mgmt-gateway
placement:
  count: 1
spec:
  port: 8888
  ssl_protocols:
  - TLSv1.3
[ceph: root@ceph-node-00 /]# 
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts]: I checked that box, if I entered values what mentione din options I could see this but If I typed out, it did not work for me.
Fpor eg. I entered "HIGH:!aNULL:!MD5;" which is default in [nginx](https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_ciphers) and it did not show up in config. 
The cipher field was absent.

Again jhghlighting, the list of ciphers provided might not be helpful here either as its rather an overwhelming cipher suites used and a combo of those.
If you try `openssl ciphers` in command shell, you would see by yourself.
IMO lets leave it to the user to type out the cipher.
All the cipher defined are seperated by a colon `:` and end with semicolon `:` as what I am seeing, so we cna put some validation as well.

Thanks

**Comment by rkachach** [comment]
> > @Pegonzal @afreen23 while testing this PR I saw that the spec generated when using default config has the `port` field set to 443:
> > ```
> > ceph orch ls --service_name mgmt-gateway --export
> > service_type: mgmt-gateway
> > service_name: mgmt-gateway
> > placement:
> >   hosts:
> >   - ceph-node-1
> > spec:
> >   port: 443
> >   ssl_protocols:
> >   - TLSv1.3 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This causes issues when generating the `redirect_uri`. I'd suggest by now to omit the port if it's the default one (443 for https and 80 for http). I'll change the cephadm code to handle this better latter, but for now I think we should adjust the dashboard side as all cephadm PRs have been already merged.
> 
> Thanks @rkachach , I have updated the PR. Now, for default:
> 
> ```
> [ceph: root@ceph-node-00 /]# ceph orch ls --service_name mgmt-gateway --export
> service_type: mgmt-gateway
> service_name: mgmt-gateway
> placement:
>   count: 1
> spec:
>   ssl_ciphers:
>   - ECDSA
>   - AES128
>   ssl_protocols:
>   - TLSv1.3
> ```
> 
> Other ports:
> 
> ```
> [ceph: root@ceph-node-00 /]# ceph orch ls --service_name mgmt-gateway --export
> service_type: mgmt-gateway
> service_name: mgmt-gateway
> placement:
>   count: 1
> spec:
>   port: 8888
>   ssl_protocols:
>   - TLSv1.3
> [ceph: root@ceph-node-00 /]# 
> ```

@Pegonzal by default (if the user has not changes the config) we shouldn't include anything in the spec section:

```
spec:
  ssl_ciphers:
  - ECDSA
  - AES128
  ssl_protocols:
  - TLSv1.3
```
Just FYI, when this field is not specified the real defaults are:

```    
ssl_ciphers              ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305;
```
And these defaults are part of the templates managed by cephadm... unfortunately there's no way at this moment for the dashboard to get these values :(

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts]: I'm not following here. Do you want to add the OpenSSL formatting for this?

As it is right now, you can select the desired ciphers, as seen in the video's description

**Comment by Pegonzal** [comment]
jenkins test api

**Review by afreen23 — APPROVED** [review]

**Comment by Pegonzal** [comment]
API failure, no problems identified: https://jenkins.ceph.com/job/ceph-api/81635/

**Comment by Pegonzal** [comment]
Rebased the PR after merging #59379 and added a [fix](https://github.com/ceph/ceph/pull/58618/files#diff-ddec8ca48d4a38889b7993eacdbe3d38a47aea7434460bfa0d67874fffea42a6R1239) to trim `allowlist_domains` , mentioned [here](https://github.com/ceph/ceph/pull/59379#issuecomment-2352467287)

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by Pegonzal** [comment]
jenkins test windows

**Comment by Pegonzal** [comment]
jenkins test windows

**Comment by rkachach** [comment]
I tested this PR (rebasing on main) and I can see that the mgmt-gateway is created correctly:

```
[ceph: root@ceph-node-0 /]# ceph orch ls --service_name mgmt-gateway --export
service_type: mgmt-gateway
service_name: mgmt-gateway
placement:
  count: 1
  hosts:
  - ceph-node-1
spec:
  enable_auth: true
```

**Pull request merged**
